### ISSUE TITLE
Regression of #53738: spurious single_use_lifetimes warning

### ISSUE THREAD

**Issue by stouset** [rust-lang/rust#69952]
Issue #53738 seems to have regressed. 

I tried this code:

```rust
#[deny(single_use_lifetimes)]

#[derive(Eq)]
struct Foo<'a, T> {
    /// a reference to the underlying secret data that will be derefed
    pub data: &'a mut T,
}
```

I expected to see this happen: No warning should be produced.

Instead, this happened:

```
error: lifetime parameter `'a` only used once
 --> src/lib.rs:4:12
  |
3 | #[derive(Eq)]
  |          -- ...is used only here
4 | struct Foo<'a, T> {
  |            ^^ this lifetime...
  |
note: lint level defined here
 --> src/lib.rs:1:8
  |
1 | #[deny(single_use_lifetimes)]
  |        ^^^^^^^^^^^^^^^^^^^^
```

### Meta

`rustc --version --verbose`:
```
rustc 1.42.0 (b8cedc004 2020-03-09)
binary: rustc
commit-hash: b8cedc00407a4c56a3bda1ed605c6fc166655447
commit-date: 2020-03-09
host: x86_64-apple-darwin
release: 1.42.0
LLVM version: 9.0
```
</details>

**Comment by jonas-schievink** [comment]
From what I can tell, the `Eq` case might have never worked, since `Debug` still doesn't warn (so the linked issue specifically maybe *didn't* regress, only a closely related example incorrectly warns). This needs confirmation though.

**Comment by cuviper** [comment]
It also fails in 1.40 and 1.41.

It gets a different error in 1.39, with `Eq` complaining that `PartialEq` is not implemented, but if you add that then it does compile successfully.

**Comment by chrissimpkins** [comment]
bisect-rustc:

searched nightlies: from nightly-2019-09-01 to nightly-2020-03-10
regressed nightly: nightly-2019-10-28
searched commits: from https://github.com/rust-lang/rust/commit/fae75cd216c481de048e4951697c8f8525669c65 to https://github.com/rust-lang/rust/commit/95f437b3cfb2fec966d7eaf69d7c2e36f9c274d1
regressed commit: https://github.com/rust-lang/rust/commit/b7176b44a203322c834302f3b515f8c10a54f2c1
source code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=2e0c8c395d4f0ae585e0ae1ad6043c6e

## Error

<details><summary>COLLAPSIBLE ERROR STACKTRACE</summary>
<p>

```bash
error: lifetime parameter `'a` only used once
 --> src/main.rs:4:12
  |
3 | #[derive(Eq)]
  |          -- ...is used only here
4 | struct Foo<'a, T> {
  |            ^^ this lifetime...
  |
note: the lint level is defined here
 --> src/main.rs:1:8
  |
1 | #[deny(single_use_lifetimes)]
  |        ^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

error: could not compile `bisectit`.

To learn more, run the command again with --verbose.
```

**Comment by spastorino** [comment]
cc @matthewjasper

**Comment by jyn514** [comment]
@dylni points out this is not actually fixed: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=3dcd3a8a7f8b85cc732d738bf6e57dc2
cc @cjgillot
